Lib/test/support/strace_helper.py

```diff
@@ -0,0 +1,173 @@
+_returncode_regex = re.compile(
+    r"\+\+\+ exited with (?P<returncode>\d+) \+\+\+")
+
+# Cached value of whether or not there is a compatible strace binary
+_strace_working: bool | None = None
+
+
+@dataclass
+class StraceEvent:
+    syscall: str
+    args: List[str]
+    returncode: str
+
+
+@dataclass
+class StraceResult:
+    strace_returncode: int
+    python_returncode: int
+    _raw_events: str
+    stdout: str
+    stderr: str
+
+    def events(self) -> List[StraceEvent]:
+        """Extract the call list information from _raw_events"""
+        matches = [
+            _syscall_regex.match(event)
+            for event in self._raw_events.splitlines()
+        ]
+        return [
+            StraceEvent(match["syscall"],
+                        [arg.strip() for arg in (match["args"].split(","))],
+                        match["returncode"]) for match in matches if match
+        ]
+
+    def sections(self) -> Dict[str:List[StraceEvent]]:
+        """Find all "MARK <X>" writes and use them to make groups of events.
+
+        This is useful to avoid variable / overhead strace events, like that
+        at interpreter startup, so a test can just check does the small case
+        under study work."""
+        current_section = "__startup"
+        sections = {current_section: []}
+        for event in self.events():
+            if event.syscall == 'write' and len(
+                    event.args) > 2 and event.args[1].startswith("\"MARK "):
+                # Found a new section, don't include the write in the section
+                # but all events until next mark should be in that section
+                current_section = event.args[1].split(
+                    " ", 1)[1].removesuffix('\\n"')
+                if current_section not in sections:
+                    sections[current_section] = list()
+            else:
+                sections[current_section].append(event)
+
+        return sections
+
+
+@support.requires_subprocess()
+def strace_python(code: str,
+                  strace_flags: List[str],
+                  check: bool = True) -> StraceResult:
+    """Run strace and return the trace.
+
+    Sets strace_returncode and python_returncode to `-1` on error
+    """
+    res = None
+
+    def _make_error(reason, details):
+        return StraceResult(
+            strace_returncode=-1,
+            python_returncode=-1,
+            _raw_events=f"error({reason},details={details}) = -1",
+            stdout=res.out if res else "",
+            stderr=res.err if res else "")
+
+    # Run strace, and get out the raw text
+    try:
+        res, cmd_line = run_python_until_end(
+            "-c",
+            textwrap.dedent(code),
+            __run_using_command=[_strace_binary] + strace_flags)
+    except OSError as err:
+        return _make_error("Caught OSError", err)
+
+    if check and res.rc:
+        res.fail(cmd_line)
+
+    # Get out program returncode
+    decoded = res.err.decode().strip()
```

Comment:
(1) Always specify a codec on decode.  BUT... (2) Instead of decoding here, it'd be better to leave output as bytes. There will be processes that do not emit valid encoded data.